assemblyscript.html.markdown

```diff
@@ -0,0 +1,275 @@
+// Implications are that a module only includes and exports concrete functions 
+// for sets of type arguments actually used and that concrete functions can be 
+// shortcutted with [static type checks] at compile time, which turned out to 
+// be quite useful.
+// Classes
+class Tuple<T1, T2> {
+  constructor(public item1: T1, public item2: T2) {
+  }
+}
+
+// Interfaces
+interface Pair<T> {
+  item1: T;
+  item2: T;
+}
+
+// And functions
+function pairToTuple <T>(p: Pair<T>): Tuple<T, T> {
+  return new Tuple(p.item1, p.item2);
+};
+
+let tuple = pairToTuple<string>({ item1: "hello", item2: "world" });
+
+// Including references to a definition file:
+/// <reference path="jquery.d.ts" />
+
+// Template Strings (strings that use backticks)
+// String Interpolation with Template Strings
+let name = 'Tyrone';
+let greeting = `Hi ${name}, how are you?`
+// Multiline Strings with Template Strings
+let multiline = `This is an example
+of a multiline string`;
+
+let numbers: Array<number> = [0, 1, 2, 3, 4];
+let moreNumbers: Array<number> = numbers;
+moreNumbers[5] = 5; // Error, elements are read-only
+moreNumbers.push(5); // Error, no push method (because it mutates array)
+moreNumbers.length = 3; // Error, length is read-only
+numbers = moreNumbers; // Error, mutating methods are missing
+
+// Tagged Union Types for modelling state that can be in one of many shapes
+type State = 
+  | { type: "loading" }
+  | { type: "success", value: i32 }
+  | { type: "error", message: string };
+
+// This is how you import external JS functions
+declare namespace console {
+  // @ts-ignore decorator
+  @external("console", "log")
+  function log(msg: string): void
+
+   // @ts-ignore decorator
+  @external("console", "error")
+  function error(msg: string): void
+}
+
+declare const state: State;
+if (state.type === "success") {
+  console.log(state.value);
+} else if (state.type === "error") {
+  console.error(state.message);
+}
+
+// Iterators and Generators
+
+// for..of statement
+// iterate over the list of values on the object being iterated
+let arrayOfAnyType = [1, "string", false];
+for (const val of arrayOfAnyType) {
+    console.log(val); // 1, "string", false
+}
+
+let list = [4, 5, 6];
+for (const i of list) {
+   console.log(i); // 4, 5, 6
+}
+
+// for..in statement
+// iterate over the list of keys on the object being iterated
+for (const i in list) {
+   console.log(i); // 0, 1, 2
+}
+
+// Type Assertion
+
+let foo = {} // Creating foo as an empty object
+foo.bar = 123 // Error: property 'bar' does not exist on `{}`
+foo.baz = 'hello world' // Error: property 'baz' does not exist on `{}`
+
+// Because the inferred type of foo is `{}` (an object with 0 properties), you 
+// are not allowed to add bar and baz to it. However with type assertion,
+// the following will pass:
+
+interface Foo { 
+  bar: i32;
+  baz: string;
+}
+
+let foo = {} as Foo; // Type assertion here
+foo.bar = 123;
+foo.baz = 'hello world'
```

Comment:
This can't be an `interface` and has to be a `class` and for the assignment after creation to work, the properties have to be optional (which doesn't work for number types) or have a default.